### PR TITLE
fix Issue 11847 - sub-pkg not available as qualified name

### PR DESCRIPTION
**Pull request by MartinNowak**
- Change start point for qualified lookups in Import (`pkg`) from
  package.d module to the wrapping Package (package module),
  thus making sub-pkgs available for qualified lookups.

This fix is preceeded by some necessary refactoring as the
wrapping package module wasn't easily accessible to Import.

Targets master as this is quite a big change and beta for 2.077.0 begins today.

### CODE REVIEW

File: src/ddmd/dmodule.d
Comment (MartinNowak) [inline src/ddmd/dmodule.d]: Mmh, looks like this now triggers for `spec/hash-map.dd`, will need to debug this.

Comment (MartinNowak) [reply src/ddmd/dmodule.d]: Forgot to check `isDocFile` in insertIntoPackageTree, should work now.

File: src/ddmd/dimport.d
Comment (WalterBright) [inline src/ddmd/dimport.d]: This is a fix for 11847, not 14446. I'm confused.

Comment (WalterBright) [reply src/ddmd/dimport.d]: ping @MartinNowak

Comment (MartinNowak) [reply src/ddmd/dimport.d]: That workaround/fix was there before, I just added the Bugzilla reference, as the comment wasn't too enlightening.

**Comment by MartinNowak** [comment]
Anything else @WalterBright?

**Comment by MartinNowak** [comment]
Ping @WalterBright

**Comment by andralex** [comment]
@MartinNowak rebase? Let's get this in.

**Comment by wilzbach** [comment]
> rebase? 

I could rebase without conflicts as only the ddmd -> dmd layout changed and git is smart enough for this.

**Comment by luismarques** [comment]
Ping @WalterBright

**Review by JinShil — APPROVED** [review]

**Comment by luismarques** [comment]
What's needed for this to be merged?

**Comment by JinShil** [comment]
> What's needed for this to be merged?

At a minimum, a rebase.  I don't generally review PRs that aren't passing.  Contributors need to be servicing their PRs periodically if they want them to move forward.

**Comment by wilzbach** [comment]
@JinShil Rebasing a PR constantly takes a lot of time and effort, which is pointless if no progress is happening. In fact, I am more than happy to rebase a PR for someone if that's the only blocking point. In this case, it looks that this is blocked on Walter's approval, so I see rebasing it as a waste of time for now :/

**Comment by JinShil** [comment]
@wilzbach, that assumes @WalterBright will direct his attention to a PR that doesn't build, and can't be tested. That seems doubtful to me.  So we end up with this catch-22 where @WalterBright is waiting for a PR that passes tests and the contributor is waiting on approval.  I don't see how that's going to work.  Someone has to break the stalemate.

**Comment by wilzbach** [comment]
I don't think that @WalterBright's attention is affected by the rebase conflict. Also as it can be seen on CircleCi it passed on the CIs before the merge conflicts have happened.
I think we need a better system here in general, where @WalterBright or @andralex approve the general direction of a PR and then delegate the further work to other reviewers.
Then as mentioned I would be glad to rebase such a PR ...

**Comment by luismarques** [comment]
@WalterBright ping?

**Comment by wilzbach** [comment]
> by the rebase conflict

BTW I did rebase this PR now and it's failing due to newly added `ModuleComponentRange` which depends/still uses `isPackageFile`.

**Comment by jacob-carlborg** [comment]
> At a minimum, a rebase. I don't generally review PRs that aren't passing. Contributors need to be servicing their PRs periodically if they want them to move forward.

@JinShil that's not really fair, especially when a PR doesn't seem to get any attention from the reviewers, [1] for example. I've already rebased that quite a few times due to all the enum changes.

[1] https://github.com/dlang/dmd/pull/7762

File: src/dmd/dmodule.d
Comment (WalterBright) [inline src/dmd/dmodule.d]: I see `pkg/package.d` and `pkg/common.d`. But the comment references `Package`. Where is this name coming from and what does it mean? What does "wrapped by 'Package'" mean?

File: src/dmd/dimport.d
Comment (WalterBright) [inline src/dmd/dimport.d]: What is a "package module"?

File: src/dmd/dmodule.d
Comment (WalterBright) [inline src/dmd/dmodule.d]: How can it return a module if it is typed as returning a Package?

Comment (WalterBright) [reply src/dmd/dmodule.d]: What is a "conflicting module"?

**Comment by luismarques** [comment]
@MartinNowak Do you need/want help to tweak your pull request based on the provided feedback?

**Comment by JinShil** [comment]
@luismarques Feel free to adopt this if you want.

**Comment by luismarques** [comment]
I've talked with Martin, he's willing to finish his work on this pull request, as long as he finds the time.

**Comment by aliak00** [comment]
ping. Is there anything stopping this other than a passing CI?

**Comment by Geod24** [comment]
Fixed in v2.091.0

**Pull request closed without merging**
